### PR TITLE
doi: handle UI for optional DOI feature

### PR DESCRIPTION
**Pull request by anikachurilova**
* closes https://github.com/CERNDocumentServer/cds-rdm/issues/163

If DOI is required: 
Screenshot 2024-12-05 at 12 28 16

If DOI is not required:
Screenshot 2024-12-05 at 12 29 07

Different options overview:
Screenshot 2024-12-05 at 12 29 36
Screenshot 2024-12-05 at 12 29 45
Screenshot 2024-12-05 at 12 30 01

Errors displaying that DOI was needed, but was not reserved (on publish of the draft):
Screenshot 2024-12-06 at 10 35 35
Screenshot 2024-12-06 at 10 35 44

### CODE REVIEW

File: invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/controls/PublishButton/PublishButton.js [hunk 1]
```diff
@@ -30,14 +32,28 @@ class PublishButtonComponent extends Component {
...
-        : DepositFormSubmitActions.PUBLISH
-    );
-    handleSubmit(event);
-    this.closeConfirmModal();
+    const { formik, raiseDOINeededButNotReserved } = this.props;
+    const noINeedOne = formik?.values?.noINeedOne;
+    if (noINeedOne && Object.keys(formik?.values?.pids).length === 0) {
+      const errors = {
+        pids: {
+          doi: "DOI is needed. Please click on the button to reserve it.",
```
Comment (kpsherva) [inline invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/controls/PublishButton/PublishButton.js:40]: shelf: translation needed (can be done later)

File: invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/api/DepositApiClient.js [hunk 2]
```diff
@@ -103,7 +103,12 @@ export class RDMDepositApiClient extends DepositApiClient {
       );
       return new DepositApiClientResponse(data, errors);
     } catch (error) {
-      const errorData = error.response.data;
+      let errorData = error.response.data;
+      const errors = this.recordSerializer.deserializeErrors(
+        error.response.data.errors || []
```
Comment (kpsherva) [inline invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/api/DepositApiClient.js:108]: why do we need the empty array?

File: invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/fields/Identifiers/PIDField.js [hunk 3]
```diff
@@ -349,11 +377,16 @@ class CustomPIDField extends Component {
...
-
+    const value = field?.value;
+    const isDraft = record?.is_draft;
     this.state = {
-      isManagedSelected: undefined,
+      isManagedSelected:
+        isDraft === true && value?.identifier && value?.provider !== PROVIDER_EXTERNAL
+          ? true
+          : undefined,
+      isNoNeedSelected: undefined,
```
Comment (kpsherva) [inline invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/fields/Identifiers/PIDField.js:389]:
```suggestion
   const isInternalProvider = value?.provider !== PROVIDER_EXTERNAL;
   const isDraft = record?.is_draft == true;
   const hasIdentifier = value?.identifier;
   const isManagedSelected =  isDraft && hasIdentifier && isInternalProvider ? true : undefined;
  
   this.state = {
      isManagedSelected: isManagedSelected;
      isNoNeedSelected: undefined,
```

also, why do we set isManagedSelected to `undefined` not `false`?

File: invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/fields/Identifiers/PIDField.js [hunk 4]
```diff
@@ -407,40 +442,72 @@ class CustomPIDField extends Component {
     }
 
     const hasManagedIdentifier = managedIdentifier !== "";
+    const hasUnmanagedIdentifier = unmanagedIdentifier !== "";
```
Comment (kpsherva) [inline invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/fields/Identifiers/PIDField.js:445]: why do we set empty string ?

Comment (zzacharo) [reply invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/fields/Identifiers/PIDField.js:445]: The initialization happens above and here we check if the value is not empty string. I am confused about the question :)

**Review by slint — APPROVED** [review]
LGTM (reviewed IRL as well)

**Pull request merged**